mirgecom/diffusion.py

```diff
@@ -0,0 +1,116 @@
+from pytools.obj_array import make_obj_array, obj_array_vectorize
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from meshmode.dof_array import thaw, DOFArray
+from grudge.symbolic.primitives import TracePair
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+
+
+def _q_flux(discr, alpha, u_tpair):
+    normal = thaw(u_tpair.int[0].array_context, discr.normal(u_tpair.dd))
+    flux_weak = math.sqrt(alpha)*u_tpair.avg*normal
+    return discr.project(u_tpair.dd, "all_faces", flux_weak)
+
+
+def _u_flux(discr, alpha, q_tpair):
+    normal = thaw(q_tpair.int[0].array_context, discr.normal(q_tpair.dd))
+    flux_weak = math.sqrt(alpha)*make_obj_array([np.dot(q_tpair.avg, normal)])
+    return discr.project(q_tpair.dd, "all_faces", flux_weak)
+
+
+def _operator(discr, alpha, w):
+    u = make_obj_array([w])
+
+    dir_u = discr.project("vol", BTAG_ALL, u)
+
+    q = discr.inverse_mass(
+        -math.sqrt(alpha)*discr.weak_grad(u[0])
+        +  # noqa: W504
+        discr.face_mass(
+            _q_flux(discr, alpha=alpha, u_tpair=interior_trace_pair(discr, u))
+            + _q_flux(discr, alpha=alpha,
+                u_tpair=TracePair(BTAG_ALL, interior=dir_u, exterior=-dir_u))
+            + sum(
+                _q_flux(discr, alpha=alpha, u_tpair=tpair)
+                for tpair in cross_rank_trace_pairs(discr, u)
+            )
+        ))
+
+    dir_q = discr.project("vol", BTAG_ALL, q)
+
+    return (
+        discr.inverse_mass(
+            make_obj_array([-math.sqrt(alpha)*discr.weak_div(q)])
+            +  # noqa: W504
+            discr.face_mass(
+                _u_flux(discr, alpha=alpha, q_tpair=interior_trace_pair(discr, q))
+                + _u_flux(discr, alpha=alpha,
+                    q_tpair=TracePair(BTAG_ALL, interior=dir_q, exterior=dir_q))
+                + sum(
+                    _u_flux(discr, alpha=alpha, q_tpair=tpair)
+                    for tpair in cross_rank_trace_pairs(discr, q))
+                )
+            )[0]
+        )
+
+
+def diffusion_operator(discr, alpha, w):
+    r"""
+    Compute the diffusion operator.
+
+    The diffusion operator is defined as
+    :math:`\nabla\cdot(\alpha\nabla w)`, where :math:`\alpha` is the diffusivity and
+    :math:`w` is a scalar field.
+
+    Parameters
+    ----------
+    discr: grudge.eager.EagerDGDiscretization
+        the discretization to use
+    alpha: float
+        the (constant) diffusivity
+    w: meshmode.dof_array.DOFArray or numpy.ndarray
+        the DOF array or object array of DOF arrays to which the operator should be
+        applied
+
+    Returns
+    -------
+    meshmode.dof_array.DOFArray or numpy.ndarray
+        the diffusion operator applied to *w*
+    """
+    if (isinstance(w, np.ndarray)
+            and w.dtype.char == "O"
+            and not isinstance(w, DOFArray)):
+        return obj_array_vectorize(lambda u: _operator(discr, alpha, u), w)
+    else:
+        return _operator(discr, alpha, w)
```

Comment:
Could you make a helper function `obj_array_vectorize_dof` in grudge that packages up this pattern, make `grudge.eager` use it, and use it here (and elsewhere in mirgecom where the pattern occurs)?

Reply:
Would it maybe make sense to add an optional condition to `obj_array_vectorize` to further limit vectorizing? Then the above could become something like

```python
    obj_array_vectorize(lambda u: _operator(discr, alpha, u), w),
                condition=lambda u: return not isinstance(u, DOFArray))
```
or
```python
    obj_array_vectorize_if(lambda u: _operator(discr, alpha, u), w),
                lambda u: return not isinstance(u, DOFArray))
```
?

Reply:
Done in https://github.com/inducer/meshmode/pull/77. I'm hoping to merge that today-ish.